### PR TITLE
Update iproto replication doc

### LINKED ISSUE #2275: New field in JOIN request:  IPROTO_SERVER_VERSION
**Product:** Tarantool
**Since:** 2.10.0-beta1
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/#box-protocol-join
**SME:** @ sergepetrenko

# Details

JOIN request now comes with a new field: replica's version.
The field uses IPROTO_SERVER_VERSION key.
Requested by @sergepetrenko in https://github.com/tarantool/tarantool/commit/528f5438dcc10bffd721c7a65e2525461f904a02.

### LINKED ISSUE #2276: New protocol stage during JOIN: IPROTO_JOIN_META, IPROTO_JOIN_SNAPSHOT
**Product:** Tarantool
**Since:** 2.10.0-beta1
**Root document:** https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/#box-protocol-join
**SME:** @ sergepetrenko

# Details

A new stage is added to the stream of JOIN rows coming from master.
The stage is marked with a bodyless row with type
IPROTO_JOIN_META = 71
Once all the rows from the stage are sent out, the JOIN continues as
before (as a stream of snapshot rows). The end of META stage is marked
with a row of type IPROTO_JOIN_SNAPSHOT = 72

The stage contains the rows that are necessary for instance
initialization (current Raft term, current state of synchronous
transaction queue), but do not belong to any system space.
Requested by @sergepetrenko in https://github.com/tarantool/tarantool/commit/f1c2127d4367fe4663b76aac4197012d5d393fd9.

Related to https://github.com/tarantool/doc/issues/2275

### LINKED ISSUE #4011: Typo in iproto > Keys used in requests and responses
> <…>ERT,
> IPROTO_REPLACE,
> IPROTO_UPSERT
> Tuple to be inserted
> 
> **IPROTO_|UPSERT|**
> Operations to perform
> 
> IPROTO_AUTH
> Array of 2 fields:
> authentic<…>

https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/keys/#internals-iproto-keys-tuple

Update

### LINKED ISSUE #4495: Complete iproto replication docs
**Doc page:** https://www.tarantool.io/en/doc/latest/reference/internals/iproto/replication/
**SME:** @ bitgorbovsky @ KAPANDR 

# Details

Add the missing info to the iproto [Replication requests and responses](https://www.tarantool.io/en/doc/latest/reference/internals/iproto/replication/) page:
- check the requests table, add missing request types if any
- IPROTO_JOIN: 
  - recheck interaction steps
  - add info about second steam: diff (new operations happened during the request handling)
  - add info about fields of _replication_ INSERT request 
- IPROTO_FETCH_SNAPSHOT: write a detailed section. In general, it's similar to IPROTO_JOIN without the diff

For details, ask @ bitgorbovsky

### PR DESCRIPTION
**Pull request by p7nov**
Resolves #4495, #2275, #2276, #4011 

Update [Reference > Internals > Binary protocol > Replication requests and responses](https://docs.d.tarantool.io/en/doc/gh-4495-iproto-replication/reference/internals/iproto/replication/):
- Add missing IPROTO_JOIN field IPROTO_SERVER_VERSION
- Add missing steps in the IPROTO_JOIN processing sequence:
  - sending instance initialization information after the first vclock (optional)
  - sending all inserts/updates/upserts/deletes after the initial insert stream is completed
- Add IPROTO_FETCH_SNAPSHOT subsection with processing sequence (similar to IPROTO_JOIN)
- Add IPROTO_JOIN_META and IPROTO_JOIN_SNAPSHOT message types
- Fix typo on [Keys used in requests and responses](https://docs.d.tarantool.io/en/doc/gh-4495-iproto-replication/reference/internals/iproto/keys/)

> [!NOTE]
> These fixes are to be cherry-picked to 2.11 documentation. 3.x iproto replication updates will be made in a separate PR.

Deployment: https://docs.d.tarantool.io/en/doc/gh-4495-iproto-replication/reference/internals/iproto/replication/

### CODE REVIEW

File: doc/reference/internals/iproto/replication.rst [hunk 1]
```diff
@@ -135,6 +163,70 @@ is an optional key used in the SUBSCRIBE request followed by an array
...
+
+IPROTO_REGISTER
+~~~~~~~~~~~~~~~
+
+Code: 0x46.
+
+To register an anonymous replica in a replica set so that it's not anonymous anymore,
+it must send an IPROTO_REGISTER request to any (TODO: or master?) node in the replica set:
+
+TODO: request structure diagram
```
Comment (p7nov) [inline doc/reference/internals/iproto/replication.rst]: @bitgorbovsky @sergepetrenko please assist with TODOs:
- which instances can accept IPROTO_REGISTER?
- what is the request structure?

File: doc/reference/internals/iproto/replication.rst [hunk 2]
```diff
@@ -42,20 +42,27 @@ General
...
-                See :ref:`replication.anon <configuration_reference_replication_anon>`
 
-        *   -   IPROTO_REGISTER
+        *   -   :ref:`IPROTO_REGISTER <box_protocol-register>`
             -   0x46
             -   Register an anonymous replica so it is not anonymous anymore
+
+        *   -   :ref:`IPROTO_JOIN_META <box_protocol-join-meta>`
+            -   0x47
+            -   A request sent in response to IPROTO_JOIN before the instance initialization information
```
Comment (sergepetrenko) [inline doc/reference/internals/iproto/replication.rst]: It's sent in response to IPROTO_JOIN or IPROTO_FETCH_SNAPSHOT
Same about IPROTO_JOIN_SNAPSHOT

File: doc/reference/internals/iproto/replication.rst [hunk 3]
```diff
@@ -135,6 +163,70 @@ is an optional key used in the SUBSCRIBE request followed by an array
...
+IPROTO_FETCH_SNAPSHOT request to any node in the replica set:
+
+..  raw:: html
+    :file: images/repl_fetch_snapshot_request.svg
+
+To learn about anonymous replicas, see :ref:`replication.anon <configuration_reference_replication_anon>`.
+
+..  include:: replication.rst
+    :start-after: iproto_join_response_sequence_start
+    :end-before: iproto_join_response_sequence_end
```
Comment (sergepetrenko) [inline doc/reference/internals/iproto/replication.rst]: A response for FETCH_SNAPSHOT is shorter than the one for JOIN:
it contains only
IPROTO_JOIN_META / IPROTO_JOIN_SNAPSHOT stage (optional)
and the IPROTO_INSERT stage (without replace,update, etc)

File: doc/reference/internals/iproto/replication.rst [hunk 4]
```diff
@@ -97,19 +104,40 @@ To join a replica set, an instance must send an initial IPROTO_JOIN request to a
...
+
+    This step applies if the IPROTO_SERVER_VERSION specified in the request is `2.10` or later.
+
+#.  A number of :ref:`INSERT <box_protocol-insert>` requests (with additional LSN and ServerID).
+    This way, the data is updated on the instance that sent the IPROTO_JOIN request.
     The instance should not reply to these INSERT requests.
 
-#.  It sends the new vclock's MP_MAP in a response similar to the one above
-    and closes the socket.
+#.  The new vclock's MP_MAP in a response similar to the one above.
```
Comment (sergepetrenko) [inline doc/reference/internals/iproto/replication.rst]: FETCH_SNAPSHOT response ends after this line.

File: doc/reference/internals/iproto/replication.rst (hunk 1 above)
Comment (sergepetrenko) [reply doc/reference/internals/iproto/replication.rst]: The request should be sent to master (like JOIN).

REGISTER contains
MP_STR instance uuid of the sender
(optional) MP_STR instance name (this only applies to 3.0 docs)
MP_MAP vclock of the instance to be registered.

A response to REGISTER contains everything from a JOIN response that is not covered by FETCH_SNAPSHOT response

Basically, IPROTO_FETCH_SNAPSHOT + IPROTO_REGISTER are equivalent to IPROTO_JOIN.

IPROTO_FETCH_SNAPSHOT performs the "initial join" stage, where a stream of snapshot rows is sent to a replica, and IPROTO_REGISTER performs the "final join" stage, where the node is registered, and a follow-up stream of xlog contents is fed to the replica (until it receives a row with its own registration).

**Review by sergepetrenko — COMMENTED** [review]
Thanks for the patch!

File: doc/reference/internals/iproto/replication.rst (hunk 2 above)
Comment (p7nov) [reply doc/reference/internals/iproto/replication.rst]: Thanks, added

**Review by sergepetrenko — APPROVED** [review]
Thanks for the fixes!
LGTM.

File: doc/reference/internals/iproto/replication.rst [hunk 5]
```diff
@@ -92,24 +101,48 @@ IPROTO_JOIN
...
+
+5.  A number of :ref:`INSERT <box_protocol-insert>`, :ref:`REPLACE <box_protocol-replace>`,
+    :ref:`UPDATE <box_protocol-update>`, :ref:`UPSERT <box_protocol-upsert>`,
+    and :ref:`DELETE <box_protocol-delete>` requests. This way, the instance
+    that is joining the replica set receives data updates that happened during
+    the join stage.
+
+6.  The new vclock's MP_MAP in a response similar to the one above.
+
+Then the instance closes the socket.
```
Comment (bitgorbovsky) [inline doc/reference/internals/iproto/replication.rst]: Соединение не закрывается, так как при поднятии репликации по тому же соединению потом посылаются запросы на SUBSCRIBE

File: doc/reference/internals/iproto/replication.rst [hunk 6]
```diff
@@ -135,6 +168,88 @@ is an optional key used in the SUBSCRIBE request followed by an array
...
+
+#.  A number of :ref:`INSERT <box_protocol-insert>`, :ref:`REPLACE <box_protocol-replace>`,
+    :ref:`UPDATE <box_protocol-update>`, :ref:`UPSERT <box_protocol-upsert>`,
+    and :ref:`DELETE <box_protocol-delete>` requests. This way, the instance
+    that is registering in the replica set receives data updates that happened
+    since the time it fetched the snapshot.
+
+#.  The new vclock's MP_MAP.
+
+Then the instance closes the socket.
```
Comment (bitgorbovsky) [inline doc/reference/internals/iproto/replication.rst]: Аналогичное замечание

File: doc/reference/internals/iproto/replication.rst [hunk 7]
```diff
@@ -135,6 +168,88 @@ is an optional key used in the SUBSCRIBE request followed by an array
...
+..  raw:: html
+    :file: images/repl_fetch_snapshot_request.svg
+
+To learn about anonymous replicas, see :ref:`replication.anon <configuration_reference_replication_anon>`.
+
+..  include:: replication.rst
+    :start-after: iproto_fetch_snapshot_response_sequence_start
+    :end-before: iproto_fetch_snapshot_response_sequence_end
+
+Then the instance closes the socket.
```
Comment (bitgorbovsky) [inline doc/reference/internals/iproto/replication.rst]: Аналогичное замечание. Откуда вообще информация, что после каждого такое взаимодействия закрывается соединение?

**Review by bitgorbovsky — CHANGES_REQUESTED** [review]

File: doc/reference/internals/iproto/replication.rst (hunk 5 above)
Comment (p7nov) [reply doc/reference/internals/iproto/replication.rst]: Спасибо, убрал это предложение из всех трёх описаний.

**Review by bitgorbovsky — APPROVED** [review]

**Pull request merged**